java/client/src/main/java/glide/connectors/handlers/CallbackDispatcher.java

```diff
@@ -58,21 +64,54 @@ public CompletableFuture<Response> registerConnection() {
     }
 
     /**
-     * Complete the corresponding client promise and free resources.
+     * Complete the corresponding client promise, handle error and free resources.
      *
      * @param response A response received
      */
     public void completeRequest(Response response) {
+        if (response.hasClosingError()) {
+            // According to https://github.com/aws/glide-for-redis/issues/851
+            // a response with a closing error may arrive with any/random callback ID (usually -1)
+            // CommandManager and ConnectionManager would close the UDS channel on ClosingException
+            responses
+                    .values()
+                    .forEach(f -> f.completeExceptionally(new ClosingException(response.getClosingError())));
+            responses.clear();
+        }
         // Complete and return the response at callbackId
         // free up the callback ID in the freeRequestIds list
         int callbackId = response.getCallbackIdx();
         CompletableFuture<Response> future = responses.remove(callbackId);
         freeRequestIds.add(callbackId);
```

Comment:
Move this to the `if (future != null)` block.  We don't want to add invalid callbackIds to freeRequestIds

Reply:
Fixed in eb537dc0.